### PR TITLE
[move-compiler] Small changes to visitors for upcoming improvements

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- Made all visitors immutable
  - Added mutable variant for typing, but the lints cant use it
- Gave the CFG to the construction of cfgir/absint visitors 
  - helpful for filtering 
  
## Test plan 

- non-functional 

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/typing/ast.rs [hunk 1]
```diff
@@ -378,6 +378,22 @@ pub fn pat(ty: Type, pat: UnannotatedPat) -> MatchPattern {
     MatchPattern { ty, pat }
 }
 
+impl ModuleCall {
```
Comment (awelc) [inline external-crates/move/crates/move-compiler/src/typing/ast.rs:381]: Is it used somewhere already or it's an addition that will be used in one of the following PRs?

**Review by awelc — APPROVED** [review]
LGTM

File: external-crates/move/crates/move-compiler/src/typing/ast.rs (hunk 1 above)
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/typing/ast.rs:381]: Following PRs

File: external-crates/move/crates/move-compiler/src/typing/match_analysis.rs [hunk 2]
```diff
@@ -42,7 +42,7 @@ struct MatchCompiler<'ctx, 'env> {
     context: &'ctx mut Context<'env>,
 }
 
-impl TypingVisitorContext for MatchCompiler<'_, '_> {
+impl TypingMutVisitorContext for MatchCompiler<'_, '_> {
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/typing/match_analysis.rs:45]: Small nit, this likely does not need to be mutable.

**Pull request merged**